mage_ai/io/google_sheets.py

```diff
@@ -0,0 +1,287 @@
+            worksheet_name (str, optional): A worksheet name to export the df. Defaults to None
+            worksheet_position (int, optional): A worksheet position to export the df. Defaults to 0
+            header_rows (int, optional): The number of rows to use as headers. Defaults to 0
+
+        Raises:
+            KeyError: To be raised if a sheet_url, sheet_id, or sheet_name is not provided
+            ValueError: To be raised if the provided worksheet does not exist.
+
+        Returns:
+            pd.DataFrame: A dataframe of the requested sheet.
+        """
+
+        worksheet = self.fetch_worksheet(
+            sheet_url=sheet_url,
+            sheet_id=sheet_id,
+            sheet_name=sheet_name,
+            worksheet_name=worksheet_name,
+            worksheet_position=worksheet_position,
+        )
+        return pd.DataFrame(worksheet.get_all_records(head=header_rows))
+
+    def export(
+        self,
+        df: pd.DataFrame,
+        sheet_url: str = None,
+        sheet_id: str = None,
+        sheet_name: str = None,
+        worksheet_name: str = None,
+        worksheet_position: int = 0,
+    ) -> None:
+        """Exports a dataframe to an existing Worksheet within a Google Sheet.
+
+        Args:
+            df (pd.DataFrame): The dataframe to be exported.
+            sheet_url (str, optional): A sheet URL to export the dataframe to. Defaults to None
+            sheet_id (str, optional): A sheet id to export the dataframe to. Defaults to None
+            sheet_name (str, optional): A sheet name (title) to export the df. Defaults to None
+            worksheet_name (str, optional): A worksheet name to export the df. Defaults to None
+            worksheet_position (int, optional): A worksheet position to export the df. Defaults to 0
+        """
+
+        worksheet = self.fetch_worksheet(
+            sheet_url=sheet_url,
+            sheet_id=sheet_id,
+            sheet_name=sheet_name,
+            worksheet_name=worksheet_name,
+            worksheet_position=worksheet_position,
+        )
+
+        with self.printer.print_msg(
+            f"Exporting dataframe to worksheet '{worksheet.title}' "
+            f"in sheet '{worksheet.spreadsheet.title}'"
+        ):
+            worksheet.update([df.columns.values.tolist()] + df.values.tolist())
+
+    def exists(
+        self,
+        sheet_url: str = None,
+        sheet_id: str = None,
+        sheet_name: str = None,
+        worksheet_name: str = None,
+        worksheet_position: int = 0,
+    ) -> bool:
+        try:
+            self.fetch_worksheet(
+                sheet_url=sheet_url,
+                sheet_id=sheet_id,
+                sheet_name=sheet_name,
+                worksheet_name=worksheet_name,
+                worksheet_position=worksheet_position,
+            )
+            return True
+
+        except Exception as ex:
+            print(ex)
```

Comment:
if we print the exception here, the exception will show up in the output for the block when running it from the notebook. not sure how useful it is to log the exception here

Reply:
Oh. Yeah it looks like a json payload and isn't great. Is there a better way to log?

Reply:
is there another way to check if the worksheet exists? it's a bit awkward to just try to fetch it and catch an exception since the worksheet could fail to be fetched for many reasons

Reply:
Yes sir!

https://stackoverflow.com/a/73385054

Reply:
![image](https://github.com/mage-ai/mage-ai/assets/59450879/3b694e68-49d2-4137-bd90-bd9c5ffaaf7d)

Ended up being a slightly different solution— looks like an APIError 404 is thrown instead